core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1045,6 +1053,66 @@ public RelBuilder snapshot(RexNode period) {
     return this;
   }
 
+
+  /**
+   * Gets column mappings of the operator.
+   *
+   * @param op operator instance
+   * @return column mappings associated with this function
+   */
+  private Set<RelColumnMapping> getColumnMappings(SqlOperator op) {
+    SqlReturnTypeInference inference = op.getReturnTypeInference();
+    if (inference instanceof TableFunctionReturnTypeInference) {
+      return ((TableFunctionReturnTypeInference) inference).getColumnMappings();
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Create a RexCall to the CURSOR function by ordinal.
+   *
+   * @param ordinal the reference to the relational input
+   * @return RexCall to CURSOR function
+   */
+  public RexNode cursor(int ordinal) {
+    return getRexBuilder().makeCall(SqlStdOperatorTable.CURSOR, getRexBuilder()
+        .makeInputRef(peek(stack.size() - ordinal - 1).getRowType(), ordinal));
+  }
+
```

Comment:
Currently the cursor function accesses the `ordinal`th element from the top of the stack. So, if we create a table function with two inputs, the left input will be cursor(1) and the right input will be cursor(0).

But we actually want them to be the other way around.

I suggest that you add an `inputCount` parameter (see: `field(inputCount, inputOrdinal, fieldOrdinal)`). Thus the left input would be `cursor(2, 0)`, and the right input would be `cursor(2, 1)`.

By the way, it is correct to call `peek` here. But `functionScan` should call pop for each input.

Reply:
Will do. But I am curious about why doing like that.